### PR TITLE
Implement the full swagger documentation for the Dlomix serving

### PR DESCRIPTION
**Pull request by dulguunb**
- Create template for the swagger.yaml
- Create reverse proxy for the triton server
- Add notes.yaml for each of the models
- Create the swagger.yaml file generation script based on the template
- Modify the dockercompose to generate the swagger.yaml and serve
-  Implement complete boilerplate code generation logic.
- Boilerplates are splitted into four main parts
  - header generation -> Generates the server/client information and library
  - input generation -> initializes the input list
  - output generation -> initializes the output list
  - result generation -> prints out the result of output layer

### CODE REVIEW

**Review by LLautenbacher — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 1]
```diff
@@ -4,6 +4,9 @@ services:
       context: .
       target: develop
     command: /models/start_triton_server.sh
+    ports:
+      - 8504:8500
+      - 8505:8501
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml:9]: if we want a request from outside for instance swagger, we would need to have an exposed port right?

[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml]: okay, just checked even without it, it works. Removed the lines

File: docker-compose.yml [hunk 2]
```diff
@@ -14,22 +17,58 @@ services:
...
     volumes:
      - .:/workspace/dlomix-serving:Z
+
+  reverse-proxy:
+    image: nginx
     depends_on:
-      serving:
-        condition: service_healthy
-    env_file:
-     - .env
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml]: I think it was a mistake that i removed, I put them back

File: docker-compose.yml [hunk 3]
```diff
@@ -14,22 +17,58 @@ services:
...
+      - 8502:8502
+    volumes:
+      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
+    command: [nginx-debug, '-g', 'daemon off;']
+
+  swagger_gen:
+    image: python
+    depends_on: 
+      - reverse-proxy
+      - serving
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml]: the depedency_on: healthy works. So I removed the pinging

File: docker-compose.yml [hunk 4]
```diff
@@ -14,22 +17,58 @@ services:
...
+    volumes:
+      - ./:/workspace
+    command:
+      - /bin/sh
+      - -c 
+      - |
+        pip install requests
+        pip install jinja2
+        pip install PyYAML
+        cd /workspace/swagger
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml]: Added two dockerfiles

File: swagger/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+`models` is a dictionary object contains everything from `v2/models/{model}/config` endpoint and `notes.yaml`
+
+One can access the specific model's notes via: ` models[i]['note'] `
+One also can access the specific model's input/output tensor via: ` models[i]['input'] ` and `models[i]['output']`
+
+
+
+# Logic of swagger_gen.py
+
+`Swagger.yaml` generation is a service that runs before serving of the Swagger. This service waits until the Triton and Reverse proxy servers are up and running. There is a logic in the script where it sends request to the reverse proxy and if it doesn't get HTTP code between 200 and 299 it assumes the service is still starting up.
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply swagger/README.md]: yeah it was possible, I added the check and removed the info from readme

File: swagger/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+## Header boilerplate
+Essentially you could generate GRPC header boiler plate you have to define the URL, and port
+`generate_grpc_header_boilierplate`
+## Ouptut boilerplate
+Then the output boilerplate where it generates a code where it appends the output tensor layers into a list.
+The input could not be generated due to you have to give the special data to the service for each model and the types of the model had to be stored somewhere to be able to generate correct input data. 
+`generate_append_output_boilerplate`
+
+# Input boilerplate
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply swagger/README.md]: fixed

File: swagger/swagger_gen.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    python_code += generate_output_boilerplate(model)
+
+    return python_code
+   
+
+def main():
+    logging.basicConfig(encoding='utf-8', level=logging.INFO)
+    with open('./model_names.json','r') as f:
+        model_names = json.load(f)
+    print(model_names[0])
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply swagger/swagger_gen.py]: changed

File: swagger/swagger_gen.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        logging.info("Waiting for serving to start: " + url)
+        try:
+            r = requests.get(url)
+            if (r.status_code >= 200 and r.status_code <= 299):
+                serving_started = True
+                logging.info("Serving started continuing the program")
+            else:
+                time.sleep(wait_time)
+        except Exception as e:
+            time.sleep(wait_time)
```
[the remark this answers is not in the document]
Comment (dulguunb) [reply swagger/swagger_gen.py]: removed the those lines

File: docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (LLautenbacher) [reply docker-compose.yml]: Not sure what you tested, the swagger service is also configured in the docker-compose file, so it is also in the docker network.

File: docker-compose.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (LLautenbacher) [reply docker-compose.yml]: You can define multiple stages in one Dockerfile and refer to them later on. You can see an example of that in the main [Dockerfile](https://github.com/wilhelm-lab/dlomix-serving/blob/main/Dockerfile). I read somewhere that this is preferred over multiple Dockerfiles let, me know if you find something that says differently!

[the remark this answers is not in the document]
Comment (dulguunb) [reply docker-compose.yml]: true, changed the code

**Comment by dulguunb** [comment]
I put the waiting logic again and exposed grpc port:
 - the healthy status does not always work so we need to have some logic that waits
 - without grpc port exposing, we would not be able to use grpc code from outside

File: swagger/swagger_gen.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dulguunb) [reply swagger/swagger_gen.py:165]: I put them back again and in separate function because health status sometimes lies

**Review by LLautenbacher — APPROVED** [review]

**Pull request merged**